### PR TITLE
[improve][ml] Reduce object allocation when iterate on LongPairRangeSet

### LINKED ISSUE #18356: [improve][broker] Reduce object allocation when iterate on `LongPairRangeSet`
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Motivation

according to #18240 

when iterate on `LongPairRangeSet` call `forEach` will:
1. convert long pair into `Range` object.
2. and then the `Range` object is passed to `LongPairConsumer<? extends T> consumer` which is `PostitionImpl::new`.

but for the case `ManagedCursorImpl.buildIndividualDeletedMessageRanges` and `ConcurrentOpenLongPairRangeSet.size`
we can get the answer by just perform on the raw (long, long) format. the object creation can be avoided.

### Solution

_No response_

### Alternatives

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by lifepuzzlefun**
Fixes #18356

### Motivation
support direct perform logic on (long,long) pair without `Range` and `PostionImpl` object creation

### Modifications
1. add `LongPairRangeSet.forEachWithRangeBoundMapper`
2. change `ManagedCursorImpl.buildIndividualDeletedMessageRanges` and `LongPairRangeSet.size` to use this function
3. add unit test to check `forEach` and `forEachWithRangeBoundMapper` return the same result.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Comment by WJL3333** [comment]
image

after jmh test. the result show this pr allocate less object and have an improved performance.

**Comment by WJL3333** [comment]
the result json can achieve here https://gist.github.com/WJL3333/bff704b2783b617fc0bc93413113e197
just upload to https://jmh.morethan.io/ get the picture.

**Comment by WJL3333** [comment]
@merlimat @codelipenghui @Technoboy- @poorbarcode PTAL

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentOpenLongPairRangeSet.java [hunk 1]
```diff
@@ -220,6 +222,35 @@ public void forEach(RangeProcessor<T> action, LongPairConsumer<? extends T> cons
         });
     }
 
+    @Override
+    public <O> void forEachWithRangeBoundMapper(LongPairConsumer<O> rawRangeBoundMapper,
```
Comment (coderzc) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentOpenLongPairRangeSet.java:226]: There is a lot of duplication of code with `forEach`, can it be optimized?

Comment (WJL3333) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentOpenLongPairRangeSet.java:226]: thanks for review. the generic for range require for `<T extands Comparable<T>>` which is used in `Range<T>` but for iteration ops we may transform to non Comparable object. i try to reuse but we need the `MLDataFormats.MessageRange` implements `Comparable`

**Comment by WJL3333** [comment]
@codelipenghui  @eolivelli  @Jason918 @lhotari @HQebupt PTAL

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 2]
```diff
@@ -2839,20 +2839,30 @@ private List<MLDataFormats.MessageRange> buildIndividualDeletedMessageRanges() {
...
-                nestedPositionBuilder.setLedgerId(p.getLedgerId());
-                nestedPositionBuilder.setEntryId(p.getEntryId());
-                messageRangeBuilder.setUpperEndpoint(nestedPositionBuilder.build());
-                MessageRange messageRange = messageRangeBuilder.build();
-                acksSerializedSize.addAndGet(messageRange.getSerializedSize());
-                rangeList.add(messageRange);
-                return rangeList.size() <= config.getMaxUnackedRangesToPersist();
-            });
+
+            individualDeletedMessages.forEachWithRangeBoundMapper(
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: Nice suggestion ! I have done coding. please take a look again.

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentOpenLongPairRangeSet.java (hunk 1 above)
Comment (WJL3333) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentOpenLongPairRangeSet.java]: Ichange the signature and optimize the code. please take a look~

**Comment by WJL3333** [comment]
All CI passed in my local repository. see https://github.com/WJL3333/pulsar/pull/2

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/LongPairRangeSet.java [hunk 3]
```diff
@@ -270,14 +305,27 @@ public void forEach(RangeProcessor<T> action) {
...
                 if (!action.process(range)) {
                     break;
                 }
             }
         }
 
+        @Override
+        public void forEachRawRange(RawRangeProcessor action) {
+            for (Range<T> range : asRanges()) {
+                LongPair lowerEndpoint = this.rangeEndPointConsumer.apply(range.lowerEndpoint());
```
Comment (315157973) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/LongPairRangeSet.java:319]: When a range does not set the lower boundary, whether there will be NPE here？

Comment (WJL3333) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/LongPairRangeSet.java:319]: thanks for review, after check the code, all range is insert by `LongPairRangeSet.addOpenClosed(long,long,long,long)` i think there is no chance to get a `Range` which boundary is not set. maybe i missed somewhere, correct me if im wrong

**Review by Technoboy- — APPROVED** [review]
LGTM

**Review by coderzc — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]

**Comment by tisonkun** [comment]
cc @codelipenghui @Technoboy- @Jason918 shall we merge this patch now?

**Pull request merged**